modules/dreamview/backend/handlers/websocket_handler.cc

```diff
@@ -46,20 +46,25 @@ void WebSocketHandler::handleReadyState(CivetServer *server, Connection *conn) {
 
 void WebSocketHandler::handleClose(CivetServer *server,
                                    const Connection *conn) {
+  // Remove from the store of currently open connections. Copy the mutex out
+  // so that it won't be reclaimed during map.erase().
+  Connection *connection = const_cast<Connection *>(conn);
+
+  std::shared_ptr<std::mutex> connection_lock;
   {
     std::unique_lock<std::mutex> lock(mutex_);
+    connection_lock = connections_[connection];
+  }
 
-    // Remove from the store of currently open connections. Copy the mutex out
-    // so that it won't be reclaimed during map.erase().
-    Connection *connection = const_cast<Connection *>(conn);
-    std::shared_ptr<std::mutex> connection_lock = connections_[connection];
-    {
-      std::unique_lock<std::mutex> lock(*connection_lock);
-      connections_.erase(connection);
-    }
-    AINFO << name_
-          << ": Connection closed. Total connections: " << connections_.size();
+  {
+    // Make sure there's no data being sent via the connection
+    std::unique_lock<std::mutex> lock_connection(*connection_lock);
+    std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
Is there a way to lock two mutexes in an atomic way?

Reply:
the other reference of this is not a scoped lock, meaning that we call ->lock, then ->unlock.